cis194/week12/zehua/test/RiskSpec.hs

```diff
@@ -0,0 +1,220 @@
+module RiskSpec where
+
+import           Control.Monad        (replicateM_)
+import           Control.Monad.Random (evalRand, evalRandIO)
+import           Data.Ix              (inRange)
+import           Risk
+import           System.Random        (RandomGen (genRange, next, split))
+import           Test.Hspec
+
+
+-- a mock to simulate random generation with a pre-defined sequence of "random" numbers
+-- we only support one single list of values, once used up _|_ will be returned
+newtype MockRandomGen = MockRandomGen { getValues :: [Int] }
+  deriving Show
+
+instance RandomGen MockRandomGen where
+  next g = case getValues g of
+             (n:ns) -> (n, MockRandomGen ns)
+             _      -> undefined
+  genRange _ = (1, 6)
+  split = undefined
+
+mkMockRandomGen :: [Int] -> MockRandomGen
+-- inserting [1,1,1,1] in between each desired random output as `random` calls `next` multiple times
+-- and it so happens that it takes 5 calls to `next` for it to be happy when genRange is (1,6)
+-- see https://hackage.haskell.org/package/random-1.1/docs/src/System.Random.html#randomIvalInteger
+mkMockRandomGen = MockRandomGen . concat . map (\x -> [1,1,1,1,x])
+
+rollsShouldBeValid :: [DieValue] -> Int -> Expectation
+rollsShouldBeValid rs n = do
+  length rs `shouldBe` n
+  -- ensure in range
+  mapM_ (\a -> a `shouldSatisfy` inRange (1, 6) . unDV ) rs
+  -- ensure sorted desc
+  mapM_ (\a -> a `shouldSatisfy` uncurry (>=)) $ zip rs (tail rs)
+
+bf :: Army -> Army -> Battlefield
+bf a d = Battlefield { attackers = a, defenders = d }
+
+
+spec :: Spec
+spec = do
+  describe "ex1" $ do
+    it "works for dieNSorted via MockRandomGen" $ do
+      (flip evalRand) (mkMockRandomGen [1,2,3,4,5,6]) $ do
+        r1 <- dieNSorted 1
+        r2 <- dieNSorted 3
+        r3 <- dieNSorted 2
+        return (\() -> do
+          r1 `shouldBe` [1]
+          r2 `shouldBe` [4,3,2]
+          r3 `shouldBe` [6,5]
+          )
```

Comment:
really could not get `Rand g a` and `Expectation` to work together. So this is the hacky way in which I managed to make the compiler agree to work 😅 

The main idea is that since I could not get the two monads to fuse together, I just run the `Rand g a` one first to produce the assertions and execution the assertions in the `IO` / `Expectation` world.